compiler/modules/front/tests/testParser.bal

```diff
@@ -0,0 +1,238 @@
+         ["E", "expr", "(", ""],
+         ["E", "expr", ")", ""],
+         ["E", "expr", "(1", ""],
+         ["V", "expr", "(v)", "v"],
+         ["E", "expr", "(1+)", ""],
+         ["V", "expr", "(f())", "f()"],
+         ["E", "expr", "(f()())", ""],
+         ["V", "expr", "(((x)))", "x"],
+         // statement
+         ["E", "stmt", "", ""],
+         ["E", "stmt", ";", ""],
+         ["E", "stmt", "1;", ""],
+         ["E", "stmt", "--a;", ""],
+         ["E", "stmt", "a + b;", ""],
+         ["V", "stmt", "break;", "break;"],
+         ["V", "stmt", "continue;", "continue;"],
+         ["V", "stmt", "a:x(c,d);", "a:x(c, d);"],
+         // statement return
+         ["V", "stmt", "return;", "return;"],
+         ["V", "stmt", "return ok;", "return ok;"],
+         ["E", "stmt", "return a, b;", "return a, b;"],
+         // statement var decl
+         ["E", "stmt", "int 1i = 0;", ""],
+         ["E", "stmt", "int i = 1(-1);", ""],
+         ["E", "stmt", "int i = (-1)1;", ""],
+         ["E", "stmt", "int i = 1 2;", ""],
+         ["E", "stmt", "(int i = 10);", ""],
+         ["V", "stmt", "int x = a != b;", "int x = a != b;"],
+         ["E", "stmt", "int i = {}", ""],
+         ["E", "stmt", "int x = a =! b;", ""],
+         ["V", "stmt", "int i = 10;", "int i = 10;"],
+         ["V", "stmt", "boolean i = 10;", "boolean i = 10;"],
+         ["V", "stmt", "boolean b = false;", "boolean b = false;"],
+         // statement assign
+         ["E", "stmt", "a = b = d;", ""],
+         ["V", "stmt", "a = 0;", "a = 0;"],
+         ["V", "stmt", "a = 0 == 1;", "a = 0 == 1;"],
+         ["V", "stmt", "a = 0 != 1;", "a = 0 != 1;"],
+         // statement if else
+         ["E", "stmt", "if a noOp(1);", ""],
+         ["E", "stmt", "if a {} else return;", ""],
+         ["E", "stmt", "if a = b {}", ""],
+         // module parts
+         ["E", "mod", "import;", ""],
+         ["U", "mod", "import x;", "import x;"],
+         ["V", "mod", "import x/y;", "import x/y;"]];
+    var cases = check file:readDir("modules/front/tests/data");
+    foreach var caseFile in cases {
+        string base = check file:basename(caseFile.absPath);
+        if !base.endsWith(SOURCE_EXTENSION) {
+            continue;
+        }
+        string case = check readCase(caseFile.absPath);
+        string parentDir = check file:parentPath(caseFile.absPath);
+        string canonCase = check file:joinPath(parentDir, canonCaseName(base));
+        string expected = case;
+        if check file:test(canonCase, file:EXISTS) {
+            expected = check readCase(canonCase);
+        }
+        string[] baseParts = regex:split(base, "-");
+        s.push([baseParts[0], baseParts[1], case, expected]);
+    }
+    return s;
+}
+
+function readCase(string path) returns string|error {
+    string[] lines = check io:fileReadLines(path);
+    string[] caseLines = [];
+    boolean inCase = false;
+    int indented = 0;
+    foreach var l in lines {
```

Comment:
I try to avoid using `l` (by itself) as a variable name: too like `I` and `1` in many fonts.